File: api/src/patients/utils/edss-calculator.ts
import { BadRequestException } from '@nestjs/common';

export interface EdssFunctionalScores {
  pyramidalSystem: number; // 0-6
  cerebellarSystem: number; // 0-5
  brainstemSystem: number; // 0-5
  sensorySystem: number; // 0-6
  bowelBladderSystem: number; // 0-6
  visualSystem: number; // 0-6
  mentalSystem: number; // 0-5
}

export interface EdssAmbulationMetrics {
  unassistedWalkingDistanceMeters?: number;
  requiresUnilateralAid?: boolean;
  requiresBilateralAid?: boolean;
  wheelchairBound?: boolean;
}

const FSS_BOUNDS: Record<keyof EdssFunctionalScores, number> = {
  pyramidalSystem: 6,
  cerebellarSystem: 5,
  brainstemSystem: 5,
  sensorySystem: 6,
  bowelBladderSystem: 6,
  visualSystem: 6,
  mentalSystem: 5,
};

export function validateFunctionalScores(scores: EdssFunctionalScores): void {
  for (const [field, max] of Object.entries(FSS_BOUNDS)) {
    const value = scores[field as keyof EdssFunctionalScores];
    if (value === undefined || value === null) {
      throw new BadRequestException(`${field} is required`);
    }
    if (!Number.isInteger(value) || value < 0 || value > max) {
      throw new BadRequestException(
        `${field} must be an integer between 0 and ${max}`,
      );
    }
  }
}

function roundToHalfStep(score: number): number {
  return Math.round(score * 2) / 2;
}

// Approximates Kurtzke's EDSS step definitions from the 7 FSS grades, for the
// band where the patient is still fully ambulatory (steps 0.0-4.5). The
// published scale is prose-based rather than a formula; ties/edge cases here
// default to the lower (less severe) step.
function scoreFromFunctionalSystems(scores: EdssFunctionalScores): number {
  const values = Object.values(scores);
  const maxGradeOfDisability = Math.max(...values);
  // Counts how many systems share a specific severity grade
  const countAt = (grade: number) => values.filter((v) => v === grade).length;

  if (maxGradeOfDisability === 0) return 0.0;
  if (maxGradeOfDisability === 1) return countAt(1) === 1 ? 1.0 : 1.5;
  if (maxGradeOfDisability === 2) {
    if (countAt(2) === 1) return 2.0;
    if (countAt(2) === 2) return 2.5;
    if (countAt(2) >= 3 && countAt(2) <= 4) return 3.0;
    return 3.5; // countAt(2) === 5
  }
  if (maxGradeOfDisability === 3) {
    if (countAt(3) === 1) return 3.0;
    if (countAt(3) === 2) return 3.5;
    return 4.0; // 3+ FS at grade 3
  }
  if (maxGradeOfDisability === 4) {
    return countAt(4) === 1 ? 4.0 : 4.5;
  }
  // maxGradeOfDisability >= 5: at least one FS severe enough that ambulation is normally
  // affected too - let the ambulation-based score take over from here.
  return 4.5;
}

// Ambulation is the primary driver for scores >= 4.0 per Kurtzke. Distance
// thresholds and aid requirements below map directly to the published step
// definitions (500m/300m unaided -> 4.0/4.5, 100m with one aid -> 6.0, 20m
// with two aids -> 6.5, wheelchair-bound -> 7.0).
//
// Note: steps 7.5-9.5 additionally depend on self-care/transfer/communication
// ability, which this model doesn't capture yet - wheelchair-bound patients
// are floored at 7.0 here and should be refined by clinician review.
function scoreFromAmbulation(ambulation: EdssAmbulationMetrics): number {
  if (ambulation.wheelchairBound) return 7.0;
  if (ambulation.requiresBilateralAid) return 6.5;
  if (ambulation.requiresUnilateralAid) return 6.0;

  const distance = ambulation.unassistedWalkingDistanceMeters;

  // If distance is missing, undefined, or effectively unrestricted (> 500m)
  if (distance === undefined || distance === null || distance > 500) {
    return 0.0;
  }

  // Evaluate from shortest (most severe) to longest (least severe) distance
  if (distance <= 100) return 5.5;
  if (distance <= 200) return 5.0; // Limited daily activity, can walk ~200m
  if (distance <= 300) return 4.5;

  // distance is between 301 and 500 meters
  return 4.0;
}

export function calculateEdssScore(
  scores: EdssFunctionalScores,
  ambulation: EdssAmbulationMetrics,
): number {
  validateFunctionalScores(scores);

  const fssScore = scoreFromFunctionalSystems(scores);
  const ambulationScore = scoreFromAmbulation(ambulation);

  return roundToHalfStep(Math.max(fssScore, ambulationScore));
}

File: calculator/edss_calculator.py
from dataclasses import dataclass
from typing import Dict, Optional


class InvalidFunctionalScoreError(ValueError):
    """Raised when functional system scores fail validation."""
    pass


@dataclass
class EdssFunctionalScores:
    pyramidal_system: int  # 0-6
    cerebellar_system: int  # 0-5
    brainstem_system: int  # 0-5
    sensory_system: int  # 0-6
    bowel_bladder_system: int  # 0-6
    visual_system: int  # 0-6
    mental_system: int  # 0-5


@dataclass
class EdssAmbulationMetrics:
    unassisted_walking_distance_meters: Optional[float] = None
    requires_unilateral_aid: Optional[bool] = False
    requires_bilateral_aid: Optional[bool] = False
    wheelchair_bound: Optional[bool] = False


FSS_BOUNDS: Dict[str, int] = {
    "pyramidal_system": 6,
    "cerebellar_system": 5,
    "brainstem_system": 5,
    "sensory_system": 6,
    "bowel_bladder_system": 6,
    "visual_system": 6,
    "mental_system": 5,
}


def validate_functional_scores(scores: EdssFunctionalScores) -> None:
    for field, max_val in FSS_BOUNDS.items():
        value = getattr(scores, field, None)

        if value is None:
            raise InvalidFunctionalScoreError(f"{field} is required")

        # Ensure value is an integer and within valid clinical bounds
        if not isinstance(value, int) or isinstance(value, bool) or not (0 <= value <= max_val):
            raise InvalidFunctionalScoreError(
                f"{field} must be an integer between 0 and {max_val}"
            )


def _round_to_half_step(score: float) -> float:
    return round(score * 2) / 2


def _score_from_functional_systems(scores: EdssFunctionalScores) -> float:
    values = [
        scores.pyramidal_system,
        scores.cerebellar_system,
        scores.brainstem_system,
        scores.sensory_system,
        scores.bowel_bladder_system,
        scores.visual_system,
        scores.mental_system,
    ]
    max_grade_of_disability = max(values)
    
    # Counts how many systems share a specific severity grade
    def count_at(grade: int) -> int:
        return values.count(grade)

    if max_grade_of_disability == 0:
        return 0.0
    if max_grade_of_disability == 1:
        return 1.0 if count_at(1) == 1 else 1.5

    if max_grade_of_disability == 2:
        count_two = count_at(2)
        if count_two == 1:
            return 2.0
        if count_two == 2:
            return 2.5
        if 3 <= count_two <= 4:
            return 3.0
        return 3.5  # count_two == 5

    if max_grade_of_disability == 3:
        count_three = count_at(3)
        if count_three == 1:
            return 3.0
        if count_three == 2:
            return 3.5
        return 4.0  # 3+ FS at grade 3

    if max_grade_of_disability == 4:
        return 4.0 if count_at(4) == 1 else 4.5

    # max_grade_of_disability >= 5: let ambulation take over from here
    return 4.5


def _score_from_ambulation(ambulation: EdssAmbulationMetrics) -> float:
    if ambulation.wheelchair_bound:
        return 7.0
    if ambulation.requires_bilateral_aid:
        return 6.5
    if ambulation.requires_unilateral_aid:
        return 6.0

    distance = ambulation.unassisted_walking_distance_meters

    # If distance is missing, None, or effectively unrestricted (> 500m)
    if distance is None or distance > 500:
        return 0.0

    # Evaluate from shortest (most severe) to longest (least severe) distance
    if distance <= 100:
        return 5.5
    if distance <= 200:
        return 5.0  # Limited daily activity, can walk ~200m
    if distance <= 300:
        return 4.5

    # distance is between 301 and 500 meters
    return 4.0


def calculate_edss_score(
    scores: EdssFunctionalScores,
    ambulation: EdssAmbulationMetrics,
) -> float:
    validate_functional_scores(scores)

    fss_score = _score_from_functional_systems(scores)
    ambulation_score = _score_from_ambulation(ambulation)

    return _round_to_half_step(max(fss_score, ambulation_score))